src/mip/utils/timestamp.hpp

```diff
@@ -0,0 +1,222 @@
+#pragma once
+
+#include <assert.h>
+#include <chrono>
+
+using std::chrono::duration_cast;
+using Nanoseconds = std::chrono::nanoseconds;
+using Microseconds = std::chrono::microseconds;
+using Milliseconds = std::chrono::milliseconds;
+using Seconds = std::chrono::seconds;
+using Minutes = std::chrono::minutes;
+using Hours = std::chrono::hours;
+
+#if _HAS_CXX20
+    using Days = std::chrono::days;
+    using Weeks = std::chrono::weeks;
+    using Months = std::chrono::months;
+    using Years = std::chrono::years;
+#else
+    #include <ratio>
+    using Days = std::chrono::duration<int, std::ratio<86400>>;
+    using Weeks = std::chrono::duration<int, std::ratio<604800>>;
+    using Months = std::chrono::duration<int, std::ratio<2629746>>;
+    using Years = std::chrono::duration<int, std::ratio<31556952>>;
+#endif // _HAS_CXX20
+
+namespace mip
+{
+    /// -----------------------------------------------------
+    /// Legend:
+    /// -----------------------------------------------------
+    /// T --> Arithmetic type
+    /// D --> Duration (currently std::chrono::duration only)
+    /// -----------------------------------------------------
+    
+
+    /// Manages a timestamp that increments based on an underlying time system.
+    ///
+    /// Notes:
+    /// ---------------------------------------------------------------------------------
+    ///     * Currently supports GPS clock (only) for the underlying time system.
+    ///     * Might add support for different time systems in the future.
+    ///     * Currently supports std::chrono:duration (only) for timestamp inputs.
+    /// ---------------------------------------------------------------------------------
+    class TimestampState
+    {
+    public:
+        TimestampState() {}
+
+        /// Synchronizes timestamp to the current time in the underlying time system.
+        void synchronize();
+
+        /// Returns time since epoch.
+        template<typename D> D getTimestamp();
+        Nanoseconds getTimestamp();
+
+    private:
+        Nanoseconds timestamp{0}; 
+    };
+
+    /// Returns time elapsed since the start of the current week of the timestamp.
+    ///
+    /// Example usage:
+    ///     // Get time of week in the same duration unit as the input timestamp.
+    ///     Nanoseconds timestamp{2000000000};
+    ///     Nanoseconds same_out = getTimeOfWeek(timestamp);
+    ///
+    ///     // Get time of week in a different duration unit from the input timestamp.
+    ///     Nanoseconds timestamp{2000000000};
+    ///     Seconds diff_out = getTimeOfWeek(duration_cast<Seconds>(timestamp));
+    template<typename D> 
+    D getTimeOfWeek(const D &timestamp);
```

Comment:
I templated the return type so that the time can be retrieved in any duration unit. This is possible since the underlying timestamp store is a duration that can be converted to any other one.

What I had before though was a second template where you manually set the return type, so something like this:
```
template<typename DReturn, typename DIn>
DReturn getTimeOfWeek(const DIn &timestamp);
```
that you would use like this:
```
Seconds test_time = getTimeOfWeek<Seconds>(Nanoseconds{500000});
```

So I could either add the set return types like you suggested, or add this template back. Which would you prefer?